wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/validator/DevfileIntegrityValidator.java

```diff
@@ -51,15 +74,42 @@
    * </ul>
    * </pre>
    *
+   * <p>Note that this doesn't validate the selectors in the devfile. If you have access to the
+   * {@link FileContentProvider} instance, you may want to also invoke {@link
+   * #validateContentReferences(Devfile, FileContentProvider)} to perform further validation that is
+   * otherwise impossible.
+   *
    * @param devfile input devfile
    * @throws DevfileFormatException if some of the checks is failed
+   * @see #validateContentReferences(Devfile, FileContentProvider)
    */
   public void validateDevfile(Devfile devfile) throws DevfileFormatException {
     validateProjects(devfile);
     Set<String> toolNames = validateTools(devfile);
     validateCommands(devfile, toolNames);
   }
 
+  /**
+   * Validates that various selectors in the devfile reference something in the referenced content.
+   *
+   * @param devfile the validated devfile
+   * @param provider the file content provider to fetch the referenced content with.
+   * @throws DevfileFormatException when some selectors don't match any objects in the referenced
```

Comment:
You wrap every exception to `DevfileFormatException`, so I think it makes sense to mention it as well. Like here https://github.com/eclipse/che/blob/master/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java#L300